src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -17,5 +21,136 @@ private Socket GetOrCreateAcceptSocket(Socket acceptSocket, bool unused, string
             handle = null;
             return null;
         }
+
+        private void SendFileInternal(string fileName, byte[] preBuffer, byte[] postBuffer, TransmitFileOptions flags)
+        {
+            // Note, UseDefaultWorkerThread is the default and is == 0.
+            // Unfortunately there is no TransmitFileOptions.None.
+            if (flags != TransmitFileOptions.UseDefaultWorkerThread)
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            // Open the file, if any
+            // Open it before we send the preBuffer so that any exception happens first
+            FileStream fileStream = null;
+            if (fileName != null && fileName.Length > 0)
+            {
+                fileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read);
+            }
+
+            SocketError errorCode = SocketError.Success;
+            using (fileStream)
+            {
+                // Send the preBuffer, if any
+                // This will throw on error
+                if (preBuffer != null && preBuffer.Length > 0)
+                {
+                    Send(preBuffer);
+                }
+
+                // Send the file, if any
+                if (fileStream != null)
+                {
+                    // This can throw ObjectDisposedException.
+                    errorCode = SocketPal.SendFile(_handle, fileStream);
+                }
+            }
+
+            if (errorCode != SocketError.Success)
+            {
+                SocketException socketException = new SocketException((int)errorCode);
+                UpdateStatusAfterSocketError(socketException);
+                if (NetEventSource.IsEnabled) NetEventSource.Error(this, socketException);
+                throw socketException;
+            }
+
+            // Send the postBuffer, if any
+            // This will throw on error
+            if (postBuffer != null && postBuffer.Length > 0)
+            {
+                Send(postBuffer);
+            }
+        }
+
+        private async Task SendFileInternalAsync(FileStream fileStream, byte[] preBuffer, byte[] postBuffer)
+        {
+            SocketError errorCode = SocketError.Success;
+            using (fileStream)
+            {
+                // Send the preBuffer, if any
+                // This will throw on error
+                if (preBuffer != null && preBuffer.Length > 0)
+                {
+                    // Using "this." seems to make the extension method kick in
+                    await this.SendAsync(new ArraySegment<byte>(preBuffer), SocketFlags.None);
```

Comment:
@jaredpar, is this by design, that there's a meaningful difference in overload resolution based on whether "this." is used?